v2/tools/mangle-test-json/main.go

```diff
@@ -0,0 +1,224 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"sort"
+	"strings"
+	"time"
+)
+
+type JSONFormat struct {
+	Time    time.Time
+	Action  string
+	Package string
+	Test    string
+	Output  string
+}
+
+type TestRun struct {
+	Action  string
+	Package string
+	Test    string
+	Output  []string
+	RunTime time.Duration
+}
+
+func main() {
+	for _, testOutputFile := range os.Args[1:] {
+		fmt.Printf("# `%s`\n\n", testOutputFile)
+
+		byPackage := loadJSON(testOutputFile)
+
+		packages := []string{}
+		for k, v := range byPackage {
+			packages = append(packages, k)
+			sort.Slice(v, func(i, j int) bool {
+				return v[i].Test < v[j].Test
+			})
+		}
+
+		sort.Strings(packages)
+
+		printSummary(packages, byPackage)
+		printDetails(packages, byPackage)
+		printSlowTests(byPackage)
+	}
+}
+
+func min(i, j int) int {
+	if i <= j {
+		return i
+	}
+
+	return j
+}
+
+func actionSymbol(d TestRun) string {
+	switch d.Action {
+	case "pass":
+		return "✅"
+	case "fail":
+		return "❌"
+	case "skip":
+		return "⏭️"
+	default:
+		panic(fmt.Sprintf("unhandled action: %s", d.Action))
+	}
+}
+
+func loadJSON(testOutputFile string) map[string][]TestRun {
+	content, err := ioutil.ReadFile(testOutputFile)
+	if err != nil {
+		log.Fatalf("%e", err)
+	}
+
+	// make test output into valid JSON
+	jsonData := "[" + strings.Join(strings.Split(strings.Trim(string(content), " \n\r"), "\n"), ",") + "]"
+
+	data := []JSONFormat{}
+	err = json.Unmarshal([]byte(jsonData), &data)
+	if err != nil {
+		log.Fatalf("%e", err)
+	}
+
+	// track when each test started running
+	startTimes := make(map[string]time.Time)
+	runTimes := make(map[string]time.Duration)
+	outputs := make(map[string][]string)
+	key := func(d JSONFormat) string {
+		return d.Package + "/" + d.Test
+	}
+
+	// package → list of tests
+	byPackage := make(map[string][]TestRun, len(data))
+	for _, d := range data {
+		if d.Action == "run" {
```

Comment:
Can this be a switch? Wondering if it would be easier to read.

Reply:
Consider yourself lucky I didn’t write this in `bash`